### PR TITLE
Battery mods

### PR DESCRIPTION
**Pull request by biell**
When you have time, please review this updated pull request.  It builds upon my previous one and brings it in-line with my changes to ponewheel.  Additionally it has placeholders to use AmpHours used and Regen'd to help make it more accurate.  My hope is that the Battery.java can be tweaked from here, without the need to change the rest of the code.

Thanks for reviewing.

### CODE REVIEW

**Comment by Nanoux** [comment]
Finally got a minute, this looks really good, thanks for making it! I'll work on putting it into the source code when I get a chance. Could you explain or link me to the algorithm your using to estimate SoC? Its hard to follow through just the code.

**Comment by biell** [comment]
Thanks.  I have a local version which is more robust, I will cancel this pull and make another one.  I wanted this in here largely so you could look through it and let me know if you were interested in it the way I have it coded.

In addition to what you see here, in the latest version I also follow the OneWheel's reported percentage against AmpHours drawn.  The "Fully Charged Two-X" version will use that if it can for a more accurate measure of battery usage.

The Voltage Output and the Voltage Cells methods follow the same algorithm, they just pull two slightly different values.  So, what follows applies to both.

1. I pull the voltage total and the Amp Hours draw (which can go negative on regen).  I apply a filter to voltage based on the draw, right now I just divide that number by about 10, which seems to work well.
2. Next, I apply a decaying average to the voltage to smooth it.  A decaying average is a way, while only keeping the previous state, of continually computing an average whereby old values are of ever decreasing priority.
3. Then, I use a function to translate the resulting voltage into a percentage.  Voltage drops off fast when a battery is full, then slowly across most of its life, then fast again when it starts to empty.  This function describes that line.
4. Finally, I apply a decaying average against that percentage to further smooth the output.

Does this make sense?  I have a google sheet I can share with you that has the steps applied and graphed, if you would like to take a look.

**Comment by biell** [comment]
Thanks, I am almost at a good stable place where this is all usable.  I am not really playing around with custom shaping, but I was moving it around because I hit the 15 notify limit.  However, that shouldn't be a problem anymore, as I added a "5th" state and am testing a timer which will pull the temperature once per minute and use that.  I noticed that the temperature change did affect my calculations.  Pulling the temperature on a standard period is more usable than having it come in only on change, that helps me smooth out its effect on range.  Any more commits will just be if I find any issues in my final testing.

File: app/src/main/java/com/nanowheel/nanoux/nanowheel/MainActivity.java [hunk 1]
```diff
@@ -341,6 +341,12 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, Strin
                 }
                 break;
 
+            case SharedPreferencesUtil.METHOD_BATTERY:
+                if (BluetoothService.isExist) {
+                    BluetoothService.mService.mOWDevice.forceBatteryRemaining();
```
Comment (biell) [inline app/src/main/java/com/nanowheel/nanoux/nanowheel/MainActivity.java]: There are different methods for battery remaining, if you change to a different method, force an update so the app reflects the new method.

File: app/src/main/java/com/nanowheel/nanoux/nanowheel/model/OWDevice.java [hunk 2]
```diff
@@ -369,11 +375,12 @@ public void setupCharacteristics(Context context) {
 
         //Read these for the diagnotics page
         deviceReadCharacteristics.add(new DeviceCharacteristic(OnewheelCharacteristicTemperature, KEY_CONTROLLER_TEMP,   "Controller Temp",4));
-        deviceReadCharacteristics.add(new DeviceCharacteristic(OnewheelCharacteristicBatteryTemp, KEY_BATTERY_TEMP,   "Battery Temperature",4));
-        deviceReadCharacteristics.add(new DeviceCharacteristic(OnewheelCharacteristicTripTotalAmpHours, KEY_TRIP_AMPS,   "Trip Amp Hours",4));
-        deviceReadCharacteristics.add(new DeviceCharacteristic(OnewheelCharacteristicTripRegenAmpHours, KEY_TRIP_AMPS_REGEN,   "Trip Amp Hours Regen",4));
         deviceReadCharacteristics.add(new DeviceCharacteristic(OnewheelCharacteristicLifetimeAmpHours, KEY_AMP_HOURS_LIFE,   "Lifetime Amp Hours",4));
 
+        //Read these once a minute
```
Comment (biell) [inline app/src/main/java/com/nanowheel/nanoux/nanowheel/model/OWDevice.java]: I created a new "state" number here (5) which will be pulled once per minute.  Right now, the only thing to pull like this is the battery temperature.  I pull it like this for two reasons: (1) the 15 notify limit is reached and (2) this is better because the temperature only updates in 5°C increments, this allows me to smooth the change out.  Without this, the battery %remaining can go up or down when the battery finally switches the temperature.

File: app/src/main/java/com/nanowheel/nanoux/nanowheel/util/BluetoothService.java [hunk 3]
```diff
@@ -47,6 +48,9 @@
     public static BluetoothService mService;
     static BluetoothUtil bluetoothUtil;
 
+    //private static Handler handler;
```
Comment (biell) [inline app/src/main/java/com/nanowheel/nanoux/nanowheel/util/BluetoothService.java]: We don't need a separate handler for for each scheduled task.  This creates one global one for Bluetooth functions.  Later, move both the Gemini and BatteryTemp functions into this single handler as postDelayed tasks.

File: app/src/main/java/com/nanowheel/nanoux/nanowheel/util/BluetoothService.java [hunk 4]
```diff
@@ -304,18 +308,31 @@ public static void updateNotification(int percent, Context context){
 
     public void deviceConnectedTimer() {
         final int repeatTime = 15000;
-        final Handler handler = new Handler();
         handler.postDelayed(new Runnable() {
             @Override
             public void run() {
-                if(isExist && SharedPreferencesUtil.getPrefs(getBaseContext()).getStatusMode() == 2) {
-                    mOWDevice.sendKeyChallengeForGemini(getBluetoothUtil());
-                    handler.postDelayed(this, repeatTime);
```
Comment (biell) [inline app/src/main/java/com/nanowheel/nanoux/nanowheel/util/BluetoothService.java]: Instead of worrying about stopping and starting the scheduled task correctly all the time (and sometimes messing up and getting two of them) lets just loop forever, but only send the key exchange when we should.

**Comment by biell** [comment]
@Nanoux this is ready to merge and have you run it through some tests.  I think it is pretty solid now.  I added a few comments for review when you read the diff to help you understand why I changed a few things that I thought you might be have questions about.

Please don't hesitate to ask questions about anything for which you are unsure of.  I had to stick my finger down into the guts in a few places.

Thank you for considering this pull request.

**Comment by biell** [comment]
Nevermind, there is still a bug somewhere.  I think I messed up in centralizing the scheduled tasks.  Back to the drawing board on that one.

**Comment by biell** [comment]
I don't think most of the stability issues were my fault.  I think I switched to a warmer jacket and my phone was further from the board, causing bluetooth connection issues.  Most of the problems were caused by that.  Regardless, it is good to have them cleaned up.

I did cause the issues fixed by making mOWDevice static.

I think this is ready for you to put it through your own tests.

**Comment by Nanoux** [comment]
I'm terribly sorry how slow I've been moving on this, life has been crazy for me. O_O

Looking through this is all looks mostly pretty good, very good work! My one question is why mOWDevice has to be static in the Bluetooth Service class? The original version of onewave I hacked together because I didn't know what I was doing had it as static, but it caused memory leaks and I did a major refactor this past summer to fix it & many others, and I would like to keep avoiding static variables as much as possible for that reason.

**Comment by biell** [comment]
I agree, and it doesn't look like this would be needed.  `new OWDevice` is only called once, and it is from `setupOWDevice`, which is in `onStartCommand`.  And so, you would think there would only ever be one OWDevice.

I implemented a constant scheduler in your code, with one `Handler` for `BluetoothService` and loop it forever, instead of trying to intelligently start and stop it.  Before, I would get more than one running simultaneously.  So, this code change fixed that.  However, it appears there are circumstances (I can't remember the exact ones) where I would get a crash due to `mOWDevice` being an invalid object.  I am sorry, but I don't remember the exact cause, but it wasn't null and it wasn't valid.  Making it static fixed those issues.  It was long enough ago now, that I can't remember exactly the circumstance.  My guess was that the `onStartcommand` was getting run a second time due to connection issues or me having to stop and restart Bluetooth on my phone, invalidating the old mOWDevice and creating a new one.

It is possible that this issue was a downstream problem of other issues which have now been fixed.  You could change it to be non-static, and do some testing to see if that problem comes back.  After I stopped getting that error message, I moved on to find other sources of crashes.

Does that make sense?

**Comment by biell** [comment]
OK, how about this new method.  I stopped using a static mOWDevice and Handler, but added two static integers.  I don't think this will cause any memory leaks.  This method will ensure that only one of each looper exists and that the most recent one is the one which exists.  So, I should not get any invalid objects.

**Comment by biell** [comment]
I just keep adding things here as I find them.  But, I am ready for this to be merged once you are OK with it.  I could work against another branch if you would prefer so you have time to review and test these changes.  Just let me know.  I am hoping we can agree on how to get this merged before I find anything else.  The latest changes are because I realized both battery temperatures change and they look like they are both valid. I figured I could report both, or average them.  I chose to report both on-screen, but average them for my Battery code.  But we can average them for display if you would prefer.

**Comment by Nanoux** [comment]
Hey, I'm terribly sorry I've been so slow with this, I'm not in that great of health and keeping quite busy. I sat down today, reinstalled linux, downloaded your pull request, pulled it up side by side with mine, and finally dug through it. A few things:

Thank you for all the tiny bug fixes, always good to have >1 mind working on these things :D

I'm not very comfortable reaching the notification characteristic limit with the app, some people use onewave because their phone supports <15 and the official app wont work for them. I was trying to get it lower than 13 and now we're back up to 15. I don't know if I'll worry about if for the next release since all version of the app are available for anyone affected by this but it is a concern of mine. It is also possible those older phones have finally all been replaced by their owners, I don't know.

Is there a reason the battery class has setter methods for the characteristics its interested in and keeping them in local variables instead fo just reading from the characteristic itself? I know figuring out how information flows in someone else's project is hard, I just wanted to know if I could fix that or if thats a workaround you need for it to work.

Finally, it makes more sense to me structurally if the battery class is a normal object stored in the OWDevice class, that lives and dies with every OWDevice instance, instead of as a nebulous static class. Is there a reason you didn't do that or can I try and make that happen?

Most of this looks very good though and I'm quite impressed! Once I get a reply from you I'll go ahead and merge it in with some tweaks on my end that I mentioned above. I also recently got a pint and will soon be giving away my +, so I also *really* need the app working on pint. Hoping to get this and pint support working and then push out a new release.

**Comment by biell** [comment]
> I'm not very comfortable reaching the notification characteristic limit with the app, some people use onewave because their phone supports <15 and the official app wont work for them. I was trying to get it lower than 13 and now we're back up to 15. I don't know if I'll worry about if for the next release since all version of the app are available for anyone affected by this but it is a concern of mine. It is also possible those older phones have finally all been replaced by their owners, I don't know.

I kinda figured as much, which is why I created the periodic characteristic reads and stopped at 15, I didn't realize you wanted to be below 15.  Could we move `OnewheelCharacteristicLightingMode` or `OnewheelCharacteristicLifetimeOdometer` down to a periodic read (once every 60 seconds) with the new state=5?  e.g. ` deviceReadCharacteristics.add(new DeviceCharacteristic(OnewheelCharacteristicLifetimeOdometer, KEY_LIFETIME_ODOMETER, "Lifetime Odometer",5));`

> Is there a reason the battery class has setter methods for the characteristics its interested in and keeping them in local variables instead fo just reading from the characteristic itself? I know figuring out how information flows in someone else's project is hard, I just wanted to know if I could fix that or if thats a workaround you need for it to work.

I was trying to keep the interface identical for Battery.java for both Onewave and pOnewheel.  I thought that would be easier to maintain as a closed package if the code diverges more between Onewave and pOnewheel, since I can't control that divergence.  If you look at both repositories, the files are identical except for the package line and some commented out Timber lines.  If changing that is necessary to get this accepted, then I can accept any changes you need to make to feel comfortable with the code.

> Finally, it makes more sense to me structurally if the battery class is a normal object stored in the OWDevice class, that lives and dies with every OWDevice instance, instead of as a nebulous static class. Is there a reason you didn't do that or can I try and make that happen?

I am not a java programmer, so I think it just seemed easier to make it static.  Also, I started this project on pOnewheel, and he used a lot of static classes, and I think I just copied it over.  That being said, making it a first class object would work for any android app, so I am all for the help to make this happen.  Thank you.

> Most of this looks very good though and I'm quite impressed! Once I get a reply from you I'll go ahead and merge it in with some tweaks on my end that I mentioned above. I also recently got a pint and will soon be giving away my +, so I also _really_ need the app working on pint. Hoping to get this and pint support working and then push out a new release.

I saw your post on kwatkins' repository.  I don't plan on getting a Pint, but I am holding out to buy whatever the XR replacement is.  I will help out making sure any Pint changes continue to work with the Battery mod code.  I am interested to see if this can evolve to help out CNR/VNR setups and I expect those to start making their way to the Pint soon.

Long story short, anything you need to do to feel comfortable enough to merge this, please feel free to do.  I will do my best to work with it going forward.

**Comment by Nanoux** [comment]
Honestly theres a lot of things that are notification characteristics that dont need to be. Ridemode and light mode only ever change if we write to them and lifetime odometer doesn't change very often. Theres some optimization to do there that I just have to find the time for.

I didn't realize you were trying to get this working in 2 projects and maintain it in both places, I dunno why I didn't pick up on that. honestly onewave and ponewheel have kind of diverged pretty far at this point and I wouldn't be suprised if they keep getting further apart. It would probably be easier if you just focused on maintaining ponewheel, and I can implement any changes you make in the modified version for onewave. My day off is coming to an end but I'll work on merging it in when I get a chance.

**Comment by Nanoux** [comment]
It just occurred to me there might be 2 battery temperatures because one might be for the battery and the other for the battery controller. I don't know that for sure or which is which tho.

**Comment by biell** [comment]
That makes sense.  Maybe one of the numbers is the average of the 4 battery sensors and the other number is the controller sensor.

**Comment by Nanoux** [comment]
alright I merged in the changes and put them on a branch here: https://github.com/Nanoux/Onewave/tree/battery_and_pint

If you could verify this all works that would be lovely, I don't have a lot of time to nor do I even know what to look for tbh.

**Comment by biell** [comment]
OK, Thank You!

I will test it.  So far I found one small copy/paste issue which needs to be resolved in `processBatteryRemaining`, I am getting more "Uh Oh!"s, and I am getting a number of dead Handler logs when the app looses connection with the OW.  I will look into these and do some tests while riding.

**Comment by Nanoux** [comment]
Ya I messed up something with the handlers, the uh ohs are from the connection timing out if you don't send a proper key challenge every 15 or so seconds ,which is what the handler did. I'll fix it when I can, apologies

**Comment by Nanoux** [comment]
Ok updated (deleted and reuploaded, I'm bad at git sorry yall) the battery & pint branch. the copy paste error is fixed and I did some work to make connections more stable. It sends a key challenge every 10 seconds instead of 15, because i noticed some variance in when the handler ran, at least on android 10, and also added a loop to keep trying to write the key challenge if the last attempt failed. Seems more stable in my initial tests so far.

**Comment by biell** [comment]
I just updated my repo and built it.  I will test over the next few days.

**Comment by biell** [comment]
I am getting a lot of disconnects.  Today I only made it to 98% battery before the app disconnected.  It takes 2 tries almost every time to get the app to pick up the onewheel.  On the first scan, it starts out with "Uh-Oh!", which I had almost completely eliminated in my branch.  I have to disable my bluetooth, then re-enable it for a scan to pick up my OW after that.  So, I will have to look and see if there is a fix in my branch that didn't get copied over.

Are you having any of these issues?

**Comment by Nanoux** [comment]
Uh, no, I've done 5-10 hours of testing and it connects every time and hasn't lost a connection once. Maybe I uploaded the wrong copy of the project, I'll check when I get home

**Comment by biell** [comment]
OK, let me know and I will clone a new tree to make sure I have a clean build environment.

**Comment by Nanoux** [comment]
Ya I uploaded the right one. In the BluetoothUtilImpl class, is the writeCharacteristic function 1 line or 4 lines? It should be 4.

**Comment by biell** [comment]
Sorry it has taken so long, it has been a busy week.  We are in store for some nice weather here and I have all week off.  So, I should get some good testing in.

Yes, I only had 1 line in my `writeCharacteristic` function.  So, clearly my testing was invalid.  I will try to get a ride in tomorrow, but will at least get some testing done on Monday.

I will let you know the results.

**Comment by biell** [comment]
So, things are, indeed, stable.  I went on two long rides to day with much success.  I only had one issue.  As I was riding, I was switching around between the different battery guestimation methods to see how they were doing.  I was having trouble selecting the correct radio button entry and I kept tilting the screen so was switching back and forth between wide-screen and tall-screen modes.  Onewave crashed once while this was happening.

I hooked my phone up to android studio when I got home, but I didn't have logs so far back as to see what happened during the crash.  It could have had to do with the redraws from swapping height and width.  It could have also been that I was holding the phone far from the board and the bluetooth connection hiccuped.

Because we are saving the battery state in a preference, I was able to start Onewave again after the crash and it picked up where it had left off, and all the calculations continued to work properly.

All-in-all, I would say you are in good shape from the Battery perspective.  Thank you so much for merging this.  I like your interface better than the official app and the pOnewheel interface.  So, it will be nice to be able to use your app with my Two-X.  Once you release this code for everyone, I think a lot of Two-X riders will want to use your app.

**Comment by Nanoux** [comment]
Thank you so much for coding it! Ive wanted this feature for a while and had no clue how to go about doing it, so I'm glad it's in and working now. I'm gonna wait till pint support is either cracked or given up on and then I'll release the update proper. If anyone wants the APK your using tho you can give it out if ya want.

**Comment by Nanoux** [comment]
Oh btw you don't have to hit the radio button itself you can tap the text next to it and it'll work as well, may be a bit easier to hit ;)
